test/e2e/external_oidc_test.go

```diff
@@ -0,0 +1,461 @@
+	require.NotEmpty(t, authResponse.AccessToken)
+	require.NotEmpty(t, authResponse.IdToken)
+
+	// ==========================================
+	// Test authentication via the kube-apiserver
+	// ==========================================
+
+	// use the OIDC id_token to do a self subject review
+	kasURL := fmt.Sprintf("%s/apis/authentication.k8s.io/v1/selfsubjectreviews", tc.kubeConfig.Host)
+	ssrReq := `{"kind":"SelfSubjectReview","apiVersion":"authentication.k8s.io/v1"}`
+	req, err := http.NewRequest(http.MethodPost, kasURL, strings.NewReader(ssrReq))
+	require.NoError(t, err)
+
+	// kubernetes uses the id_token to identify the user (instead of the access_token)
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", authResponse.IdToken))
+	req.Header.Set("Content-Type", "application/json")
+	resp, err = httpClient.Do(req)
+	require.NoError(t, err)
+	defer resp.Body.Close()
+	require.Equal(t, http.StatusCreated, resp.StatusCode)
+
+	body, err := io.ReadAll(resp.Body)
+	require.NoError(t, err)
+	var ssr authenticationv1.SelfSubjectReview
+	err = json.Unmarshal(body, &ssr)
+	require.NoError(t, err)
+	require.Equal(t, email, ssr.Status.UserInfo.Username)
+	require.Contains(t, ssr.Status.UserInfo.Groups, "system:authenticated")
+}
+
+func newTestClient(t *testing.T) (*testClient, error) {
+	tc := &testClient{
+		t:          t,
+		kubeConfig: test.NewClientConfigForTest(t),
+	}
+
+	var err error
+	tc.kubeClient, err = kubernetes.NewForConfig(tc.kubeConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	tc.configClient, err = configclient.NewForConfig(tc.kubeConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	tc.operatorConfigClient, err = operatorversionedclient.NewForConfig(tc.kubeConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	return tc, nil
+}
+
+func (tc *testClient) setupKeycloakClient(ctx context.Context, keycloakURL string) *test.KeycloakClient {
+	transport, err := rest.TransportFor(tc.kubeConfig)
+	require.NoError(tc.t, err)
+
+	kcClient := test.KeycloakClientFor(tc.t, transport, keycloakURL, "master")
+	err = wait.PollUntilContextTimeout(ctx, 5*time.Second, 30*time.Second, true, func(ctx context.Context) (bool, error) {
+		err := kcClient.AuthenticatePassword(oidcClientId, "", "admin", "password")
+		if err != nil {
+			tc.t.Logf("failed to authenticate to Keycloak: %v", err)
+			return false, nil
+		}
+		return true, nil
+	})
+	require.NoError(tc.t, err)
+
+	return kcClient
+}
+
+func (tc *testClient) setupExternalOIDCWithKeycloak(ctx context.Context) (kcClient *test.KeycloakClient, cleanups []func()) {
+	kcClient, idpName, c := test.AddKeycloakIDP(tc.t, tc.kubeConfig, true)
+	cleanups = append(cleanups, c...)
+
+	// update the authentication CR with the external OIDC configuration
+	authConfig, c, err := tc.updateAuthForOIDC(ctx, kcClient.IssuerURL(), idpName)
+	cleanups = append(cleanups, c...)
+	require.NoError(tc.t, err)
+	require.NotNil(tc.t, authConfig)
+	require.NotEmpty(tc.t, authConfig.Spec.OIDCProviders)
```

Comment:
One stupid question : ): in case one `require.XXX` assertion fails, will the `cleanups` still be executed in a later time?

Reply:
`require` uses `t.FailNow()` under the hood, which in turn calls `runtime.Goexit()` -- deferred funcs will still be executed after these are called.

From https://pkg.go.dev/runtime#Goexit:
> Goexit runs all deferred calls before terminating the goroutine.